packages/notifications/push/amplify_push_notifications/android/src/main/kotlin/com/amazonaws/amplify/amplify_push_notifications/AmplifyPushNotificationsPlugin.kt

```diff
@@ -3,110 +3,263 @@
+
+
+    /**
+     * Main engine's binary messenger.
+     */
+    private var mainBinaryMessenger: BinaryMessenger? = null
+
+    /**
+     * Flutter API interface.
+     */
+    private var flutterApi: PushNotificationsFlutterApi? = null
+
+    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        mainBinaryMessenger = flutterPluginBinding.binaryMessenger
+        PushNotificationsHostApi.setup(mainBinaryMessenger, this)
+        flutterApi = PushNotificationsFlutterApi(mainBinaryMessenger)
+        applicationContext = flutterPluginBinding.applicationContext
+        sharedPreferences =
+            applicationContext.getSharedPreferences(
+                PushNotificationPluginConstants.SHARED_PREFERENCES_KEY,
+                Context.MODE_PRIVATE
+            )
+    }
+
+    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
+        flutterApi = null
     }
 
     override fun onAttachedToActivity(binding: ActivityPluginBinding) {
+        mainActivity = binding.activity
         activityBinding = binding
-        onNewIntent(binding.activity.intent)
+        mainBinaryMessenger?.let {
+            StreamHandlers.initStreamHandlers()
+            StreamHandlers.initEventChannels(it)
+        }
         binding.addOnNewIntentListener(this)
-
-        // TODO: also fetchToken on app resume
+        onNewIntent(binding.activity.intent)
         refreshToken()
-
     }
 
     override fun onDetachedFromActivityForConfigChanges() {
         activityBinding?.removeOnNewIntentListener(this)
+        activityBinding = null
+        mainActivity = null
     }
 
     override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
+        mainActivity = binding.activity
         activityBinding = binding
+        mainBinaryMessenger?.let {
+            StreamHandlers.initStreamHandlers()
+            StreamHandlers.initEventChannels(it)
+        }
         binding.addOnNewIntentListener(this)
     }
 
+
     override fun onDetachedFromActivity() {
         activityBinding?.removeOnNewIntentListener(this)
+        mainActivity = null
         activityBinding = null
     }
 
-    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
-        StreamHandlers.initialize(flutterPluginBinding.binaryMessenger)
+    // TODO(Samaritan1011001): 1. This gets called with intent only when a specific notification is tapped and not when the group is tapped
+    //      2. The intent here is the last notification device got rather than the one that was tapped
+    override fun onNewIntent(intent: Intent): Boolean {
+        intent.extras?.let {
+            val payload = it.asPayload()
+            if (payload != null) {
+                val notificationHashMap = payload.asChannelMap()
+                flutterApi?.onLaunchNotificationOpened(
+                    notificationHashMap as Map<Any, Any?>
```

Comment:
Added the cast since this was the exception case of sending the notification via `flutterApi` and conforming to it's required type as opposed to `eventChannel` but I guess it does make sense to change type for `asChannelMap`. So made that change.